### ISSUE TITLE
Invalid inference for NoReturn type of async def

### ISSUE THREAD

**Issue by tamuhey** [microsoft/pyright#670]
```python
from typing import Awaitable, NoReturn

async def foo():
    while True:
        pass

a: Awaitable[NoReturn] = foo()
```

pyright output:

```
  9:26 - error: Expression of type "NoReturn" cannot be assigned to declared type "Awaitable[NoReturn]"
  "__await__" is not present (reportGeneralTypeIssues)
```

`foo` actually return `coroutine object`, so it is `Awaitable`

### version info

- pyright 1.1.36

**Comment by erictraut** [comment]
Thanks for the bug report.

I'm not sure what the correct behavior is in this case. I special-cased NoReturn for Awaitable objects because I don't know what it means to wait for something that never returns. What type is returned by `await foo()` in your example?

I referred back to the PEPs and mypy, and they don't provide any clarity here. In fact, it appears that mypy's support for coroutines with an inferred return type is much more limited than Pyright's.

Since I don't have a strong opinion about this particular case and you appear to have an expectation that it should be typed as `Awaitable[NoReturn]`, I'm OK removing my special-case check for NoReturn and wrap it in an Awaitable for coroutines.

I'm curious whether this case has a practical use.

This change will be in the next release of Pyright.

**Comment by tamuhey** [comment]
Thanks for fixing!

I created `foo` as an observer, and it is used as follows:

```python
loop = asyncio.get_running_loop()
task = loop.create_task(foo()) // Pyright rejects this line now due to type mismatch
```

So `await[NoReturn]` is useful when it is concurrently used to observe other tasks.

**Comment by erictraut** [comment]
This is now addressed in Pyright 1.1.37, which I just published.
